src/mpl2/src/SACoreHardMacro.cpp

```diff
@@ -0,0 +1,300 @@
+{
+  flip_prob_ = flip_prob;
+}
+
+float SACoreHardMacro::CalNormCost()
+{
+  float cost = 0.0;  // Initialize cost
+  if (norm_area_penalty_ > 0.0)
+    cost += area_weight_ * (width_ * height_) / norm_area_penalty_;
+  if (norm_outline_penalty_ > 0.0)
+    cost += outline_weight_ * outline_penalty_ / norm_outline_penalty_;
+  if (norm_wirelength_ > 0.0)
+    cost += wirelength_weight_ * wirelength_ / norm_wirelength_;
+  if (norm_guidance_penalty_ > 0.0)
+    cost += guidance_weight_ * guidance_penalty_ / norm_guidance_penalty_;
+  if (norm_fence_penalty_ > 0.0)
+    cost += fence_weight_ * fence_penalty_ / norm_fence_penalty_;
+  return cost;
+}
+
+void SACoreHardMacro::CalPenalty()
+{
+  CalOutlinePenalty();
+  CalWirelength();
+  CalGuidancePenalty();
+  CalFencePenalty();
+}
+
+void SACoreHardMacro::FlipMacro()
+{
+  /*
+  macro_id_ = static_cast<int>(std::floor(
+              (distribution_)(generator_) * macros_.size()));
+  const float prob = (distribution_) (generator_);
+  macros_[macro_id_].Flip(false);
+
+  if (prob <= 0.5)
+    macros_[macro_id_].Flip(true);
+  else
+    macros_[macro_id_].Flip(false);
+  */
+  for (auto& macro : macros_)
+    macro.Flip(false);
+}
+
+void SACoreHardMacro::Perturb()
+{
+  if (macros_.size() == 0)
+    return;
+
+  // Keep back up
+  pre_pos_seq_ = pos_seq_;
+  pre_neg_seq_ = neg_seq_;
+  pre_width_ = width_;
+  pre_height_ = height_;
+  pre_outline_penalty_ = outline_penalty_;
+  pre_wirelength_ = wirelength_;
+  pre_guidance_penalty_ = guidance_penalty_;
+  pre_fence_penalty_ = fence_penalty_;
+
+  // generate random number (0 - 1) to determine actions
+  const float op = (distribution_)(generator_);
+  const float action_prob_1 = pos_swap_prob_;
+  const float action_prob_2 = action_prob_1 + neg_swap_prob_;
+  const float action_prob_3 = action_prob_2 + double_swap_prob_;
+  const float action_prob_4 = action_prob_3 + exchange_prob_;
+  const float action_prob_5 = action_prob_4 + flip_prob_;
+  if (op <= action_prob_1) {
+    action_id_ = 1;
+    SingleSeqSwap(true);  // Swap two macros in pos_seq_
+  } else if (op <= action_prob_2) {
+    action_id_ = 2;
+    SingleSeqSwap(false);  // Swap two macros in neg_seq_;
+  } else if (op <= action_prob_3) {
+    action_id_ = 3;
+    DoubleSeqSwap();  // Swap two macros in pos_seq_ and
+                      // other two macros in neg_seq_
+  } else if (op <= action_prob_4) {
+    action_id_ = 4;
+    ExchangeMacros();  // exchange two macros in the sequence pair
+  } else {
+    action_id_ = 5;
+    pre_macros_ = macros_;
+    FlipMacro();  // Flip one macro
+  }
+
+  // update the macro locations based on Sequence Pair
+  PackFloorplan();
+  // Update all the penalties
+  CalPenalty();
+  if (action_id_ == 105)
+    std::cout << "wirelength_weight_ = " << wirelength_weight_ << "  "
+              << "pre_wirelength_ = " << pre_wirelength_ << "  "
+              << "wirelength_ = " << wirelength_ << "   " << std::endl;
```

Comment:
remove

Reply:
There are still many stdout -- that are is require for debugging the test cases. After we validate the QoR of all the CI + MacroPlacement examples redundant couts will be deleted and relevant debug msgs will go as logger reports